### ISSUE TITLE
Test all paired fields' CheckConstraints

### ISSUE THREAD

**Issue by ghickman** [opensafely-core/job-server#3996]
We have various fields that are paired up, such that they should either both be null or neither should be null, typically fields such as `created_at` and `created_by` or adjacent fields.

While demoing these to the team we realised they are not all tested.  Coverage doesn't help us here because the CheckConstraints will always fire when a model is used.

An example of these tests can be found in [the Repo model tests](https://github.com/opensafely-core/job-server/blob/main/tests/unit/jobserver/models/test_repo.py#L9-L25), where we test:
1. both set
2. neither set
3. we get the expected IntegrityError with each of the relevant fields not set while the other is set

Note: the final test must use the `django_db(transaction=True)` mark so that it uses [TransactionTestCase](https://docs.djangoproject.com/en/5.0/topics/testing/tools/#django.test.TransactionTestCase) under the hood.

We'll need to look at each model that uses CheckConstraints and confirm there are tests for them.

**Comment by ghickman** [comment]
There are some extra nuances here that I'd forgotten about but hit in #4009.

When a pair of fields are both required then we don't need a constraint, so this likely only affects a subset of our fields.

**Comment by iaindillingham** [comment]
Couple of questions to help me understand, @ghickman 🙂 

- Why don't we decorate the both set/neither set tests with `pytest.mark.django_db`? In other words, how is it that the check constraints on `Repo` are applied, without requiring the database? (I can see that these tests fail when I modify them. But I don't understand why they fail.)
- I'd like to probe your preference for two tests rather than one parameterized test. I guess that's a matter of style.

**Comment by iaindillingham** [comment]
> When a pair of fields are both required then we don't need a constraint

Is a field required when `null=False` (the default)? As it looks like `created_at` and `created_by` are often required, then does this mean we can remove the associated check constraints and tests?

**Comment by ghickman** [comment]
> Why don't we decorate the both set/neither set tests with pytest.mark.django_db?

We started doing that for all tests, as it's the default for pytest-django to require opting into database access, but after a while we realised nearly all tests needed it, so we [enabled it globally](https://github.com/opensafely-core/job-server/blob/22dcf6fbf5526d688dfe1014156decafed081bee/tests/conftest.py#L56-L58).  The tests which fail the constraint then need to be marked with `transaction=True` so that error is bubbled up to our test code.

> I'd like to probe your preference for two tests rather than one parameterized test. I guess that's a matter of style?

I actually can't remember if I tried this!  For the "both set" and "neither set" tests, I can imagine that might be easier, but I'm always cautious of drying up tests as readability and clear, explicit setup is very useful!  For the test which deals with IntegrityError that can't be combined with the other type as it's a different base test class.

> Is a field required when null=False (the default)? As it looks like created_at and created_by are often required, then does this mean we can remove the associated check constraints and tests?

That's, exactly it 👍  I realised this while moving Project's created_by and updated_by to be required.  The constraints are great, but not needing to use them is even better 🎉
